cosmos/x/evm/plugins/precompile/plugin.go

```diff
@@ -119,30 +120,54 @@ func (p *plugin) SetTransientKVGasConfig(transientKVGasConfig storetypes.GasConf
 // Run implements core.PrecompilePlugin.
 func (p *plugin) Run(
 	evm ethprecompile.EVM, pc vm.PrecompileContainer, input []byte,
-	caller common.Address, value *big.Int, suppliedGas uint64, readonly bool,
+	caller common.Address, value *big.Int, suppliedGas uint64, readOnly bool,
 ) ([]byte, uint64, error) {
 	// use a precompile-specific gas meter for dynamic consumption
 	gm := storetypes.NewInfiniteGasMeter()
 	// consume static gas from RequiredGas
 	gm.ConsumeGas(pc.RequiredGas(input), "RequiredGas")
 
-	// get native Cosmos SDK context from the Polaris StateDB
+	// get native Cosmos SDK context and MultiStore from the Polaris StateDB
 	sdb := utils.MustGetAs[vm.PolarisStateDB](evm.GetStateDB())
 	ctx := sdk.UnwrapSDKContext(sdb.GetContext())
+	ms := utils.MustGetAs[MultiStore](ctx.MultiStore())
+
+	// make sure the readOnly is only set if we aren't in readOnly yet, which also makes sure that
+	// the readOnly flag isn't removed for child calls (taken from geth core/vm/interepreter.go)
+	if readOnly && !ms.IsReadOnly() {
+		ms.SetReadOnly(true)
+		defer func() { ms.SetReadOnly(false) }()
+	}
 
 	// disable reentrancy into the EVM
 	p.disableReentrancy(sdb)
 
-	// run precompile container
-	ret, err := pc.Run(
+	// recover from any panic during precompile execution for the EVM to handle as a vm error
+	var err error
+	defer func() {
+		if panicked := recover(); panicked != nil {
+			// NOTE: this only propagates an error back to the EVM if the type of the given panic
+			// value is error or string (any other type of panic value is ignored)
+			var isError bool
+			if err, isError = utils.GetAs[error](panicked); !isError {
+				if strErr, isString := utils.GetAs[string](panicked); isString {
+					err = errors.New(strErr)
+				}
+			}
+		}
+	}()
```

Comment:
Also allows any arbitrary panic in precompiles to be caught here so that
- method handler does not crash
- tx gets included, but will have a vm error (also more descriptive for devs)
- gas actually gets consumed for a bad precompile call (causing a panic)